explore/marginal-gain/1.Marginal-Gain-Multiple-Mutations.py

```diff
@@ -0,0 +1,253 @@
+from sklearn.linear_model import SGDClassifier
+from sklearn.model_selection import train_test_split, GridSearchCV
+from sklearn.metrics import roc_auc_score, roc_curve
+from sklearn.pipeline import Pipeline, FeatureUnion
+from sklearn.preprocessing import StandardScaler, Imputer, FunctionTransformer
+from sklearn.feature_selection import SelectKBest
+from statsmodels.robust.scale import mad
+from IPython.display import display
+import gc
+
+
+# In[2]:
+
+get_ipython().magic('matplotlib inline')
+plt.style.use('seaborn-notebook')
+
+
+# ## Load Data
+
+# In[3]:
+
+get_ipython().run_cell_magic('time', '', "path = os.path.join('..', '..', 'download', 'covariates.tsv')\ncovariates = pd.read_table(path, index_col=0)")
+
+
+# In[4]:
+
+get_ipython().run_cell_magic('time', '', "path = os.path.join('..', '..', 'download', 'expression-matrix.tsv.bz2')\nexpression = pd.read_table(path, index_col=0)")
+
+
+# In[5]:
+
+get_ipython().run_cell_magic('time', '', "path = os.path.join('..','..','download', 'mutation-matrix.tsv.bz2')\nY = pd.read_table(path, index_col=0)")
+
+
+# In[6]:
+
+X = {}
+X['model a'] = covariates
+X['model b'] = pd.concat([covariates,expression], axis=1)
+for k in ['model a','model b']:
+    print(X[k].shape)
+
+
+# In[80]:
+
+mutations = {
+    '7157': 'TP53',   # tumor protein p53
+    '7428': 'VHL',    # von Hippel-Lindau tumor suppressor
+    '29126': 'CD274', # CD274 molecule
+    '672': 'BRCA1',   # BRCA1, DNA repair associated
+    '675': 'BRCA2',   # BRCA2, DNA repair associated
+    '238': 'ALK',     # anaplastic lymphoma receptor tyrosine kinase
+    '4221': 'MEN1',   # menin 1
+    '5979': 'RET',    # ret proto-oncogene
+}
+
+
+# ## Median absolute deviation feature selection
+
+# In[81]:
+
+def fs_mad(x, y):
+    """    
+    Get the median absolute deviation (MAD) for each column of x
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+
+# ## Define pipeline and Cross validation model fitting
+
+# In[82]:
+
+# Parameter Sweep for Hyperparameters
+
+param_grid = {
+    'classify__loss': ['log'],
+    'classify__penalty': ['elasticnet'],
+    'classify__alpha': [1],#10.0 ** np.linspace(-3, 1, 10),
+    'classify__l1_ratio': [0.15],
+}
+
+expression_feats = Pipeline(steps=[
+    ('dim_red', FunctionTransformer(lambda X: X[:,covariates.shape[1]:])),
+    ('select', SelectKBest(fs_mad,2000)),
+])
+
+covariate_feats = Pipeline(steps=[
+    ('dim_red', FunctionTransformer(lambda X: X[:,:covariates.shape[1]])),
+])
+
+combo_pipeline = Pipeline([
+    ('imputer', Imputer()),
+    ('standardize', StandardScaler()),
+    ('features', FeatureUnion([
+        ('covariates_feats', covariate_feats),           
+        ('expression_feats', expression_feats)
+    ])),
+    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+])
+
+covariates_pipeline = Pipeline(steps=[
+    ('imputer', Imputer()),
+    ('standardize', StandardScaler()),
+    ('select', SelectKBest(fs_mad,'all')),
```

Comment:
I assume you just mean the ('select', SelectKBest(fs_mad,'all')) step. I will take it out (and modify how the coef_df dataframe is built to handle lack of 'select' step).